urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+            + aDiff.text + postpatch_text.substring(char_count2);
+        break;
+      case DELETE:
+        patch.length1 += aDiff.text.length();
+        patch.diffs.add(aDiff);
+        postpatch_text = postpatch_text.substring(0, char_count2)
+            + postpatch_text.substring(char_count2 + aDiff.text.length());
+        break;
+      case EQUAL:
+        if (aDiff.text.length() <= 2 * Patch_Margin
+            && !patch.diffs.isEmpty() && aDiff != diffs.getLast()) {
+          // Small equality inside a patch.
+          patch.diffs.add(aDiff);
+          patch.length1 += aDiff.text.length();
+          patch.length2 += aDiff.text.length();
+        }
+
+        if (aDiff.text.length() >= 2 * Patch_Margin) {
+          // Time for a new patch.
+          if (!patch.diffs.isEmpty()) {
+            patch_addContext(patch, prepatch_text);
+            patches.add(patch);
+            patch = new Patch();
+            // Unlike Unidiff, our patch lists have a rolling context.
+            // http://code.google.com/p/google-diff-match-patch/wiki/Unidiff
+            // Update prepatch text & pos to reflect the application of the
+            // just completed patch.
+            prepatch_text = postpatch_text;
+            char_count1 = char_count2;
+          }
+        }
+        break;
+      }
+
+      // Update the current character count.
+      if (aDiff.operation != Operation.INSERT) {
+        char_count1 += aDiff.text.length();
+      }
+      if (aDiff.operation != Operation.DELETE) {
+        char_count2 += aDiff.text.length();
+      }
+    }
+    // Pick up the leftover patch if not empty.
+    if (!patch.diffs.isEmpty()) {
+      patch_addContext(patch, prepatch_text);
+      patches.add(patch);
+    }
+
+    return patches;
+  }
+
+  /**
+   * Given an array of patches, return another array that is identical.
+   * @param patches Array of Patch objects.
+   * @return Array of Patch objects.
+   */
+  public LinkedList<Patch> patch_deepCopy(LinkedList<Patch> patches) {
+    LinkedList<Patch> patchesCopy = new LinkedList<Patch>();
+    for (Patch aPatch : patches) {
+      Patch patchCopy = new Patch();
+      for (Diff aDiff : aPatch.diffs) {
+        Diff diffCopy = new Diff(aDiff.operation, aDiff.text);
+        patchCopy.diffs.add(diffCopy);
+      }
+      patchCopy.start1 = aPatch.start1;
+      patchCopy.start2 = aPatch.start2;
+      patchCopy.length1 = aPatch.length1;
+      patchCopy.length2 = aPatch.length2;
+      patchesCopy.add(patchCopy);
+    }
+    return patchesCopy;
+  }
+
+  /**
+   * Merge a set of patches onto the text.  Return a patched text, as well
+   * as an array of true/false values indicating which patches were applied.
+   * @param patches Array of Patch objects
+   * @param text Old text.
+   * @return Two element Object array, containing the new text and an array of
+   *      boolean values.
+   */
+  public Object[] patch_apply(LinkedList<Patch> patches, String text) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.